src/sentry.rs

```diff
@@ -83,3 +93,147 @@ fn tracing_setup() -> Option<tracing_appender::non_blocking::WorkerGuard> {
+                }
+            }
+        }
+    }
+}
+
+fn tracing_gc() -> TracingGCGuard {
+    use std::{
+        io::{Error, ErrorKind, Result, SeekFrom},
+        path::{Path, PathBuf},
+    };
+    use tokio::{fs::File, io::AsyncSeekExt};
+
+    const GC_THRESHOLD: usize = 10_000_000;
+    const GC_THREADS: usize = 1;
+
+    /// Scan for logging files on disk.
+    #[tracing::instrument(err, skip(dir))]
+    async fn read_dir(dir: impl AsRef<Path>) -> Result<Vec<PathBuf>> {
+        let mut entries = tokio::fs::read_dir(dir).await?;
+        let mut children = Vec::new();
+        while let Some(entry) = entries.next_entry().await? {
+            let entry_name = entry.file_name();
+            let entry_name = entry_name.to_string_lossy();
+            if !entry_name.starts_with(LOG_FILENAME) {
+                continue;
+            }
+            children.push(entry.path());
+        }
+        Ok(children)
+    }
+
+    /// Fetch logging file sizes.
+    #[tracing::instrument(err, skip(paths))]
+    async fn size_of(paths: &[impl AsRef<Path>]) -> Result<Vec<usize>> {
+        let mut sizes = Vec::new();
+        for path in paths {
+            let mut file = File::open(path).await?;
+            let pos = file.seek(SeekFrom::End(0)).await?;
+            sizes.push(pos as usize);
+        }
+        Ok(sizes)
+    }
+
+    /// Return first index where the accumulation exceeds gc threshold.
+    /// For example:
+    /// - with threshold=100 and sizes=[200, 10, 10], it will return 0,
+    /// - with threshold=100 and sizes=[70, 30, 50], it will return 1,
+    /// - with threshold=100 and sizes=[60, 30, 40], it will return 2,
+    /// - with threshold=100 and sizes=[10, 20, 30], it will return 3,
+    fn find_garbage(sizes: &[usize]) -> usize {
+        let mut acc = 0;
+        for (index, size) in sizes.iter().enumerate() {
+            acc += size;
+            if acc >= GC_THRESHOLD {
+                return index;
+            }
+        }
+        sizes.len()
+    }
+
+    #[tracing::instrument(err, skip(paths))]
+    async fn erase_all(paths: &[impl AsRef<Path>]) -> Result<()> {
+        let mut last_error = Ok(());
+        for child in paths {
+            let child = child.as_ref();
+            last_error = tokio::fs::remove_file(child).await;
+            tracing::debug!("GCed {child:?}: {last_error:?}");
+        }
+        last_error
+    }
+
+    #[tracing::instrument(err)]
+    async fn attempt_gc() -> Result<()> {
+        let gc_dir = log_dir().ok_or(Error::from(ErrorKind::NotFound))?;
+
+        let mut files = read_dir(gc_dir).await?;
+
+        // sort files by descending order: newer files first.
+        files.sort_by(|lhs, rhs| lhs.file_name().cmp(&rhs.file_name()).reverse());
+
+        let sizes = size_of(&files[..]).await?;
+
+        let gc_index = find_garbage(&sizes[..]);
+
+        let to_collect = files.iter().skip(gc_index).collect::<Vec<_>>();
+        erase_all(&to_collect[..]).await
```

Comment:
i think you may run into the possibility of removing the current logging file or?

Reply:
yes, is that alright?

Reply:
what happens if you remove it while you're writing to it?

Reply:
Ah yeah either the unlink or write syscall will emit an error. I suppose the GC task is fast enough to avoid such error, and I sure hope tracing is smart enough to re-open the file, but I guess we could avoid these kind of errors altogether. Let me push a commit for this 😉